### PR TITLE
Change how existing vocabularies are loaded in preprocessing

### PR DESCRIPTION
**Pull request by goncalomcorreia**
This PR addresses the issue #945 (and possibly #989?). Instead of ignoring tokens that are not present in the current corpus, it keeps all the tokens referenced in the vocab path and also in the specified order of said file.

A new script is also added to the `tools` directory - `create_vocabulary.py`. This script allows to take a previously preprocessed corpus' `*.vocab.pt` file or a corpus text file (or several) and create a vocabulary file to use in the preprocessing stage.

### CODE REVIEW

File: onmt/inputters/inputter.py [hunk 1]
```diff
@@ -304,7 +304,20 @@ def build_vocab(train_dataset_files, fields, data_type, share_vocab,
 
     # Load vocabulary
     src_vocab = load_vocabulary(src_vocab_path, tag="source")
+    if src_vocab is not None:
+        src_vocab_size = len(src_vocab)
+        logger.info('Loaded source vocab has %d tokens.' % src_vocab_size)
+        for i, token in enumerate(src_vocab):
+            # keep the order of tokens specified in the vocab file by
+            # adding them to the counter with decreasing counting values
+            counter['src'][token] = src_vocab_size - i
```
Comment (pltrdy) [inline onmt/inputters/inputter.py]: It seems a bit arbitrary to me. Could you elaborate why you think it's necessary/relevant?

File: onmt/inputters/inputter.py [hunk 2]
```diff
@@ -378,7 +391,7 @@ def load_vocabulary(vocabulary_path, tag=""):
     """
     vocabulary = None
     if vocabulary_path:
-        vocabulary = set([])
+        vocabulary = []
```
Comment (pltrdy) [inline onmt/inputters/inputter.py]: What is the reason to use a `list` instead of a `set`?

File: onmt/inputters/inputter.py (hunk 1 above)
Comment (goncalomcorreia) [reply onmt/inputters/inputter.py]: If we only count the words in the corpus that are present in the vocabulary file we will end up with a vocabulary in the Counter object that not only may be smaller than the vocab in the specified file but also that is in a completely different order. Keeping the order and size of a specified vocabulary file can be useful, for instance, in semi-supervised settings (e.g. fusion techniques - https://arxiv.org/pdf/1503.03535.pdf) where the output softmax of a model needs to be aligned with the output softmax of a pretrained language model.

What I did there in the code is a hack, of course. I'm artificially giving descending counts to the tokens in the vocabulary list so that when torchtext's Vocab is created the order of the specified vocabulary file is maintained.

File: onmt/inputters/inputter.py (hunk 2 above)
Comment (goncalomcorreia) [reply onmt/inputters/inputter.py]: Same reason as above - to make sure the vocabulary is in the same sorting order as the specified file.

**Comment by vince62s** [comment]
thanks merging.
